File: Anytime_RRT_Sharp/helper/generate_test_scenarios.m

clear all
close all
clc

numObstacles = ceil(rand * 50 + 5);

xMin = 0;
xMax = 150;
yMin = 0;
yMax = 150;
zMin = 0;
zMax = 150;
rMin = 1;
rMax = 10;

x = zeros(1, numObstacles);
y = zeros(1, numObstacles);
z = zeros(1, numObstacles);
r = zeros(1, numObstacles);

xRange = xMax - xMin;
yRange = yMax - yMin;
zRange = zMax - zMin;
rRange = rMax - rMin;

figure()
hold on

for i = 1:numObstacles
   x(i) = xMin + rand * xRange;
   y(i) = yMin + rand * yRange;
   z(i) = zMin + rand * zRange;
   r(i) = rMin + rand * rRange;
   spherePlot(x(i), y(i), z(i), r(i))
end

[goalX, goalY, goalZ] = generatePoint(xMin, xMax, yMin, yMax, zMin, zMax, x, y, z, r);
[startX, startY, startZ] = generatePoint(xMin, xMax, yMin, yMax, zMin, zMax, x, y, z, r);

plot3(goalX, goalY, goalZ, 'g*');
plot3(startX, startY, startZ, 'b*');

hold off

outputData(startX, startY, startZ, goalX, goalY, goalZ, x, y, z, r)

function [x, y, z] = generatePoint(xMin, xMax, yMin, yMax, zMin, zMax, obsX, obsY, obsZ, obsR)
    xRange = xMax - xMin;
    yRange = yMax - yMin;
    zRange = zMax - zMin;
    
    x = xMin + rand * xRange;
    y = yMin + rand * yRange;
    z = zMin + rand * zRange;

    while (inObstacle(x, y, z, obsX, obsY, obsZ, obsR))
        x = xMin + rand * xRange;
        y = yMin + rand * yRange;
        z = zMin + rand * zRange;
    end
end

function inObstacle = inObstacle(x, y, z, obsX, obsY, obsZ, obsR)
    inObstacle = false;
    num = size(obsX,2);
    for i = 1:num
        xDiff = (x - obsX(i))^2;
        yDiff = (y - obsY(i))^2;
        zDiff = (z - obsZ(i))^2;
        dist = sqrt(xDiff + yDiff + zDiff);
        if (dist <= obsR(i))
            inObstacle = true;
            break;
        end
    end
end

function outputData(startX, startY, startZ, goalX, goalY, goalZ, obsX, obsY, obsZ, obsR)
    t = datetime('now','TimeZone','utc','Format','yyyyMMddHHmmss');
    folder = "testData_" + string(t);
    mkdir(folder);
    
    obsInfo = [];
    numObs = size(obsX, 2);
    for i = 1:numObs
        obsInfo = [obsInfo, "SPHERE", obsX(i), obsY(i), obsZ(i), obsR(i)];
    end

    obsFormat = "%s %f %f %f %f\n";
    obsFile = fullfile(folder, "obstacles.txt");
    writelines("FORMAT: (SPHERE x y z radius) (RECTANGLE minX minY minZ maxX maxY maxZ)", obsFile);
    fid = fopen(obsFile, "a+");
    fprintf(fid, obsFormat, obsInfo);
    fclose(fid);

    stateFormat = "%f, %f, %f, %f\n";
    stateFile = fullfile(folder, "states.txt");
    writelines("FORMAT: (startX, startY, startZ, startTheta) (goalX, goalY, goalZ, goalTheta)", stateFile);
    fid = fopen(stateFile, "a+");
    fprintf(fid, stateFormat, [startX, startY, startZ, 0], [goalX, goalY, goalZ, 0]);
    fclose(fid);
end

File: Anytime_RRT_Sharp/helper/generate_test_scenarios.py
import os
from datetime import datetime
from random import randrange
from basic_geometry import Point, Rectangle, Sphere
from plotting_tools import generatePlot

GOAL_RADIUS = 2.5

def generatePointAvoidingObstacles(xRange: tuple, yRange: tuple, zRange: tuple, pitchRangeDeg: tuple, yawRangeDeg: tuple, sphereObs: list[Sphere], rectanObs: list[Rectangle]) -> Point:
    point = Point.generateRandom(xRange, yRange, zRange, pitchRangeDeg, yawRangeDeg)
    while (any(x.intersects(point) for x in sphereObs) or any(x.intersects(point) for x in rectanObs)):
        point = Point.generateRandom(xRange, yRange, zRange, pitchRangeDeg, yawRangeDeg)
    return point

def writeOutputData(start: Point, goal: Point, spheres: list[Sphere], rectangles: list[Rectangle]) -> None:
    t = datetime.utcnow().strftime("%Y%m%d%H%M%S")
    folder = "testData_" + t
    os.mkdir(folder)

    obsFile = os.path.join(folder, "obstacles.txt")
    stateFile = os.path.join(folder, "states.txt")

    obsFile = open(obsFile, "w+")
    stateFile = open(stateFile, "w+")

    obsFile.write("FORMAT: (SPHERE x y z radius) (RECTANGLE minX minY minZ maxX maxY maxZ)\n")
    for s in spheres:
        obsFile.write("SPHERE {0} {1} {2} {3}\n".format(s.x, s.y, s.z, s.r))
    for r in rectangles:
        obsFile.write("RECTANGLE {0} {1} {2} {3} {4} {5}\n".format(r.minPoint.x, r.minPoint.y, r.minPoint.z, r.maxPoint.x, r.maxPoint.y, r.maxPoint.z))
    obsFile.close()

    stateFile.write("FORMAT: (startX startY startZ startPitch startYaw) (goalX goalY goalZ goalTheta goalYaw goalRadius)\n")
    stateFile.write("{0} {1} {2} {3} {4}\n".format(start.x, start.y, start.z, start.pitch, start.yaw))
    stateFile.write("{0} {1} {2} {3} {4} {5}\n".format(goal.x, goal.y, goal.z, goal.pitch, goal.yaw, GOAL_RADIUS))
    stateFile.close()

### define ranges for data generation ###
numSphereObsRange = (5, 20)
numRectanObsRange = (5, 30)
xRange = (0, 750)
yRange = (0, 750)
sphereObsZRange = (40, 120)
sphereObsRadiusRange = (1, 10)
rectanObsZRange = (0, 0)
rectanObsHeightRange = (10, 80)
rectanObsWidthRange = (10, 30)
startZRange = (150, 200)
goalZRange = (0, 5)
pitchRangeDeg = (-5, 5)
yawRangeDeg = (0, 360)
#########################################

numSphereObs = randrange(numSphereObsRange[0], numSphereObsRange[1])
numRectanObs = randrange(numRectanObsRange[0], numRectanObsRange[1])

sphereObs = [Sphere.generateRandom(xRange, yRange, sphereObsZRange, sphereObsRadiusRange) for _ in range(numSphereObs)]
rectanObs = [Rectangle.generateRandom(xRange, yRange, rectanObsZRange, rectanObsWidthRange, rectanObsHeightRange) for _ in range(numRectanObs)]

startPoint = generatePointAvoidingObstacles(xRange, yRange, startZRange, pitchRangeDeg, yawRangeDeg, sphereObs, rectanObs)
goalPoint = generatePointAvoidingObstacles(xRange, yRange, goalZRange, pitchRangeDeg, yawRangeDeg, sphereObs, rectanObs)

writeOutputData(startPoint, goalPoint, sphereObs, rectanObs)
generatePlot(startPoint, goalPoint, GOAL_RADIUS, sphereObs, rectanObs)